harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+    - [UNIMPLEMENTED] something to batch Timings and periodically flush to sender_thread (timings_batcher)
+
+    The ProfilerAgent needs to be created at the beginning of training and it needs
+    to be notified every time the batch_idx increases.
+
+    You can also ship Timings through the ProfilerAgent with the record_timing() method. This
+    functionality has not yet been implemented - we need to batch the timings and periodically
+    flush them to the sender_thread.
+
+    Profiling is only active between start_on_batch and end_after_batch. It will also automatically
+    shut down 5 minutes after starting. When profiling is not active, no system metrics are collected
+    and the record_timing function is a no-op.
+
+    Usage:
+    ```
+    profiler_agent = ProfilerAgent(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch)
+
+    for batch_idx, batch in enumerate(batches):
+        profiler_agent.update_batch_idx(batch_idx)
+
+        # NOTE: Timing API has not been fully developed yet
+        forward_pass_timing = Timing("forward_pass")
+        forward_pass_timing.start()
+        # Do forward pass
+        forward_pass_timing.end()
+        profiler_agent.record_timing(forward_pass_timing)
+    ```
+    """
+    def __init__(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch):
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Track duration to stop collecting after 5 minutes. start_time also serves as
+        # the indicator that collection has begun.
+        self.start_time = None
+        # TODO: Currently we only check for shutdown in update_batch_idx(). This is a problem when there
+        #       is a long period of time between batch_idx being updated (such as during validation)
+        #       because it won't shut down until validation completes. We probably need to create a
+        #       new thread to enforce the timeout correctly.
+        self.max_collection_seconds = 300
+        self.has_finished = False
+
+        # Set up the thread responsible for making API calls
+        self.send_queue = queue.Queue()
+        self.sender_thread = ProfilerSenderThread(self.send_queue, self.master_url)
+        self.sender_thread.start()
+
+        # Launch the system metric collecting thread, but not the actual collection of metrics
+        self.sys_metric_collector_thread = SysMetricCollectorThread(trial_id, agent_id, self.send_queue)
+        self.sys_metric_collector_thread.start()
+
+        # TODO: Add data structure to batch timings and then send to SenderThread
+        #       Does this need to be its own thread to flush correctly?
+        # self.timings_batcher = TimingsBatcher()
```

Comment:
as far as I see, this could nearly be the same code as the `SysMetricBatcher`, if it either it was smart enough to know which names go to which type or `add_nongpu_measurement` took a type too. All the rest of the code looks like it would work fine.